jwst/straylight/src/calc_xart.c

```diff
@@ -0,0 +1,273 @@
+encoded within a reference file; this function uses those parameters in conjunction
+with the observed detector image to model the cross-artifact contribution.
+
+The output of this function is a single array: (xart_flux) giving the 1d detector
+representation of the cross-artifact model.
+
+Parameters
+----------
+imin : int
+   Starting column to fit (1/2 detector at a time)
+imax : int
+   Ending column to fit
+xsize : int
+   X axis detector size
+ysize : int
+   Y axis detector size
+xvec : double array
+   Vector of X pixel values across the detector.
+fimg : double array
+   Detector flux values mapped to a single 1d array.
+gamma : double array
+   Gamma parameter of the Lorenztian for each detector row.
+lor_amp : double array
+   Amplitude of the Lorenztian for each detector row.
+g_std : double array
+   Standard deviation width of the gaussians for each detector row
+   1x for inner gaussian pair, 2x for outer gaussian pair.
+g_dx : double array
+   Linear offset of the gaussians for each detector row
+   1x for inner gaussian pair, 2x for outer gaussian pair.
+g1_amp : double array
+    Amplitude of the inner gaussians for each detector row.
+g2_amp : double array
+    Amplitude of the outer gaussians for each detector row.
+
+Returns
+-------
+xart_flux : numpy.ndarray
+  1d cross-artifact detector model
+*/
+
+#include <stdlib.h>
+#include <math.h>
+#include <stdio.h>
+#include <Python.h>
+#include <stdbool.h>
+#include <numpy/arrayobject.h>
+#include <numpy/npy_math.h>
+
+#define PY_ARRAY_UNIQUE_SYMBOL _jwst_xart_numpy_api
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+
+//_______________________________________________________________________
+// Allocate the memory for the output array, nelem elements long
+//_______________________________________________________________________
+
+int alloc_xart_arrays(int nelem, double **fluxv) {
+
+    const char *msg = "Couldn't allocate memory for output arrays.";
+
+    // flux:
+    if (!(*fluxv  = (double*)calloc(nelem, sizeof(double)))) {
+        PyErr_SetString(PyExc_MemoryError, msg);
+        goto failed_mem_alloc;
+    }
+
+    return 0;
+
+ failed_mem_alloc:
+    free(*fluxv);
+   
+}
+
+// Do the computation of the cross-artifact values.
+// return values: xart_flux
+int xart_model(int imin, int imax, int xsize_det, int ysize_det,
+		 double *xvec,
+		 double *fimg, double *gamma, double *lor_amp,
+		 double *g_std, double *g_dx, double *g1_amp, double *g2_amp,
+		 double **xart_flux) {
+
+  double *fluxv;  // vector for xart values
+  int i, j, k;
+
+  // This is how big the output vector needs to be
+  int npt = xsize_det * ysize_det;
+
+  // allocate memory to hold output 
+  if (alloc_xart_arrays(npt, &fluxv)) return 1;
+
+  for (j = 0; j < ysize_det; j++) {
+    for (i = imin; i < imax; i++) {
+      for (k = 0; k < xsize_det; k++) {
+        fluxv[xsize_det * j + k] += (fimg[j * xsize_det + i] * lor_amp[j] * gamma[j] * gamma[j])
```

Comment:
Question: the array fluxv is allocated memory for 'npt' elements =xsize_det * ysize_det . But in the for loop it looks like fluxv will contain 'xsize_det * ysize_det + imax' elements.  I am sure you have tested this and if I was correct it would give you an error - so I must be missing something.

Reply:
I think this is ok; fluxv has size xsize_det * ysize_det.  It gets populated into element xsize_det*j + k

The maximum value of j is ysize_det-1 and the maximum value of k is xsize_det -1

So the maximum element should be xsize_det * (ysize_det-1) + xsize_det - 1 = xsize_det*ysize_det - xsize_det + xsize_det - 1 = xsize_det * ysize_det - 1

Likewise since imax is never bigger than 1031=xsize_det-1

j*xsize_det + i can never be more than (ysize_det-1)*xsize_det + xsize_det-1 = xsize*det*ysize_det - xsize_det + xsize_det -1 = xsize_det*ysize_det - 1

Hasn't crashed in running it pretty extensively.